### DISCUSSION TITLE
Insert and remove SD-Cards without reboot

### DISCUSSION THREAD

**Discussion by WolleKette** [micropython/micropython#11049] [General]
First of all: Is it possible in Micropython to insert SD-Cards, read files from it and remove the card after all?

I use a WEMOS D32 Pro which has a sd-card on board. 
(But this is just for the beginning to test. Later, a separate SD card module is to be used. )

When I insert a card for the first time after power on, then I can mount the card as following:
```py
import os, machine
sd=machine.SDCard(slot=2)
sd.info().  # => (31457280000, 512)

os.mount(sd, '/sd')
...
os.umount('/sd')
```

But when I remove and reinsert the card, it's no longer possible to access the module.
No matter what I have tried, I only get error messages. 
Creating an SDCard object again for example, brings an OSError: (-259, 'ESP_ERR_INVALID_STATE')

**Comment by vascolow** [comment]
yes, same problem with me, feel so sorry to our dearest host and comities... Still learning how to split out 2 directory like what u mention @WolleKette

**Reply by WolleKette** [reply]
In the meantime I had another problem with the SD card access, but I found a solution. This also solves the problem described here. I have published it in a separate discussion. You can see if it helps you. 

It is here: [Strange behaviour of the SDCard Module #11323](https://github.com/orgs/micropython/discussions/11323)

**Comment by shariltumin** [comment]
You can try the following sequence when swapping cards using machine.SDCard().

1. ```os.umount('/sd')```
2. ```sd.deinit()```
3. ```del sd```
4. - remove card
5. ```sd=machine.SDCard(slot=2)```
6. - Insert new card
7. ```os.mount(sd, '/sd')```

The following is the log of a test run on my REPL


```
MicroPython v1.20.0-328-g01c758e26-dirty on 2023-08-03; ESP32 module with ESP32
>>> import os, machine
>>> sd=machine.SDCard(slot=2)
>>> sd=machine.SDCard(slot=2)
I (6203080) gpio: GPIO[5]| InputEn: 0| OutputEn: 1| OpenDrain: 0| Pullup: 0| Pulldown: 0| Intr:0 
>>> os.mount(sd, '/sd')
I (6319780) sdspi_transaction: cmd=5, R1 response: command not supported
>>> os.listdir('/sd')
[]
>>> sd.info()
(31666995200, 512)
>>> os.umount('/sd')
>>> sd.deinit()
I (6365280) gpio: GPIO[5]| InputEn: 1| OutputEn: 0| OpenDrain: 0| Pullup: 0| Pulldown: 0| Intr:0 
I (6365280) gpio: GPIO[23]| InputEn: 0| OutputEn: 0| OpenDrain: 0| Pullup: 1| Pulldown: 0| Intr:0 
I (6365290) gpio: GPIO[19]| InputEn: 0| OutputEn: 0| OpenDrain: 0| Pullup: 1| Pulldown: 0| Intr:0 
I (6365300) gpio: GPIO[18]| InputEn: 0| OutputEn: 0| OpenDrain: 0| Pullup: 1| Pulldown: 0| Intr:0 
>>> del sd
>>> # no card inserted
>>> sd=machine.SDCard(slot=2)
I (6465010) gpio: GPIO[5]| InputEn: 0| OutputEn: 1| OpenDrain: 0| Pullup: 0| Pulldown: 0| Intr:0 
>>> # insert card
>>> os.mount(sd, '/sd')
I (6491790) sdspi_transaction: cmd=5, R1 response: command not supported
>>> os.listdir('/sd')
['data.log']
>>> sd.info()
(8053063680, 512)
```

Your REPL output will probably not have the debugging information as above. I am using a custom firmware with debugging. As you can see, ```sd.deinit()``` will deinit the underlying GPIOs. The 1.st sd card is 32GB and the 2.nd card is 8GB.
